integration/windows_device_test.go

```diff
@@ -0,0 +1,92 @@
+//go:build windows
+// +build windows
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+func TestWindowsDevice(t *testing.T) {
+	testPodLogDir, err := os.MkdirTemp("", "windows-device")
+	require.NoError(t, err)
+	defer os.RemoveAll(testPodLogDir)
+
+	t.Log("Create a sandbox with log directory")
+	sb, sbConfig := PodSandboxConfigWithCleanup(t, "sandbox", "windows-device",
+		WithPodLogDirectory(testPodLogDir),
+	)
+
+	var (
+		// TODO: An image with the device-dumper
+		testImage     = GetImage(BusyBox)
+		containerName = "test-container"
+	)
+
+	const (
+		// Mount this device class to expose host-GPU-accelerated DirectX inside the container
+		// https://techcommunity.microsoft.com/t5/containers/bringing-gpu-acceleration-to-windows-containers/ba-p/393939
+		GUID_DEVINTERFACE_DISPLAY_ADAPTER = "5B45201D-F2F2-4F3B-85BB-30FF1F953599" //nolint:revive
+	)
+
+	EnsureImageExists(t, testImage)
+
+	t.Log("Create a container to run the device test")
+	cnConfig := ContainerConfig(
+		containerName,
+		testImage,
+		// Per C:\windows\System32\containers\devices.def, enabling GUID_DEVINTERFACE_DISPLAY_ADAPTER
+		// will mount the host driver store into the container at this location.
+		WithCommand("sh", "-c", "ls -d /Windows/System32/HostDriverStore/* | grep /Windows/System32/HostDriverStore/FileRepository"),
```

Comment:
I guess this is fine as a test, but given the info here https://github.com/kubernetes/kubernetes/issues/97739#issuecomment-767932810, I don't really feel comfortable relying on the behavior of `devices.def` in general. I don't know of plans to remove it or anything like that, but this leaks the low level implementation details of HCS that aren't documented anywhere.

Reply:
That's fair, but I'm not sure at this point how else to verify device parameter has been processed correctly.

I know the Kubernetes tests for this functionality [have had issues identifying if the device was activated or not](https://github.com/kubernetes/kubernetes/pull/93948#issuecomment-726790812), so I'm not keen to rely on my originally-planned approach of querying DXGI since I think I have less commitment from various containerd developers or CI platforms running the testsuite to have real hardware GPUs available in their environments, than I do for `C:\windows\system32\containers\devices.def` to not change behaviour on (at least) released versions of Windows. ^_^

Perhaps there's some _other_ kind of device class I could use instead, that would be 100% distinguishable and doesn't rely on actual hardware characteristics. I haven't really looked that hard, but it might be worth exploring.

Reply:
Do you know if `pnputil` or `devcon` are available in this test image?

Reply:
It does not appear to have either of those, no. The image in this case is `k8s.gcr.io/e2e-test-images/busybox:1.29-2` which I guess is nanoserver-based (just from the size), but I imagine it wouldn't be a huge problem to build another image that is nanoserver plus a particular tool (even a custom-written one like dxdumper is) if there's a particular API that could be queried.

I imagine nanoserver+PowerShell might be workable in the short-term as well, if the distinction is visible via PowerShell.

Reply:
I was thinking something like https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/pnputil-command-syntax using `/enum-devices <class guid>`. Is using a different image an option that the maintainers would be okay with?

Reply:
I imagine a different image would be okay, but it'd have to be small, so I guess nanoserver with a copy of pnputil.exe from the relevant server core image would work. Although I don't know how we want to manage/authenticate/store images used in this test, the current images are mostly inherited from critest or k8s e2e tests, AFAIK.

I'm also interested to know if you have a specific device class in-mind for this. `pnputil` doesn't seem to enumerate anything for `5B45201D-F2F2-4F3B-85BB-30FF1F953599`, and I don't know off-hand what device or device class would make a good test. (And device-activation in containers on my personal machine seems to be blue-screening about 1 time in 4, so trying to find one by brute force is a slow process.)

Reply:
I think the issue there is that you're trying to use a [device interface class guid](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/overview-of-device-interface-classes) when pnputil is looking for the [class guid](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/overview-of-device-setup-classes). (Why they chose to make that so confusing, we'll never know 😁). But that might mean the option of using `pnputil /enum-devices /class` is not really useful here. (Sorry about that, my fault). 

Although it's less nice, you could instead do `pnputil /enum-devices /interfaces` and parse the results for the target device interface class guid. I think this would work for the interface class guid you have now but I'm not 100% sure. 

But since this PR was merged, I'm okay punting on this for now if that's preferred.

Reply:
You're right, `pnputil /enum-interfaces /class "{5B45201D-F2F2-4F3B-85BB-30FF1F953599}"` does indeed return results. The remaining problem is to find an interface class where the devices are reliably distinguishable even on random test-host VMs. My concern with this one in particular is that the test-host VM might only have the "Microsoft Basic Render Driver" itself, and hence success is indistinguishable.

I'm happy to punt this for now, particularly to see how things evolve with `devices.def` over time, in case an opportunity presents itself for a more-reliable test.

That said, I probably should have add a test covering the negative case too; I might slip that in to my next piece of work, making the Base Spec in the config work correctly for Windows.